README.md
SMAA for R
==========

[![Build Status](https://travis-ci.com/gertvv/rsmaa.svg?branch=master)](https://travis-ci.com/gertvv/rsmaa)

An R package for the Stochastic Multi-criteria Acceptability Analysis (SMAA)
family of Multiple Criteria Decision Analysis (MCDA) methods.

Integration tests
-----------------

The `itest` directory contains examples that are too time-consuming or take
too much space to submit to CRAN.


smaa/NEWS
smaa 0.3-0
==========

Released 2018-05-21

 * Add pair-wise winning indices (smaa.pwi)

 * Register native code

 * Fix typos in documentation


smaa 0.2-5
==========

Released 2016-12-23

 * Update maintainer email

smaa 0.2-4
==========

Released 2015-07-09

 * Explicitly import functions from R core packages

smaa 0.2-3
==========

Released 2014-12-03

 * Use .Call interface for all C code

smaa 0.2-2
==========

Released 2013-12-02

 * Correctly use C99 "inline" facility

smaa 0.2-1
==========

Released 2013-11-01

 * Make smaa.cf preserve dimnames

 * Bugfix: check for 0 in smaa.entropy.choice

 * Implement smaa.entropy.ranking in C for 40x performance increase

smaa 0.2
==========

Released 2013-08-22

 * Add convenience method to efficiently calculate piece-wise linear partial
   value functions.

 * Fix central weights in case alternative obtains first rank in exactly one
   iteration.

 * Fix confidence factors when central weights are NA

smaa 0.1-1
==========

Released 2013-07-19

 * Fix Windows build.

smaa 0.1
==========

Released 2013-07-18

 * Initial release. Computation of alternative values, rankings, rank
   acceptabilities, central weights, confidence factors, decision entropy.
   Appropriate summaries and plots.


smaa/tests/pvf.R
library('smaa')

pvf1 <- function(x) {
  smaa.pvf(x, cutoffs=c(-0.15, 0.35), values=c(0, 1))
}
pvf2 <- function(x) {
  smaa.pvf(x, cutoffs=c(50, 100), values=c(1, 0))
}


stopifnot(all.equal(pvf1(0.35), 1))
stopifnot(all.equal(pvf1(-0.15), 0))
stopifnot(all.equal(pvf1(0.1), 0.5))
stopifnot(all.equal(pvf1(c(0.35, 0.1)), c(1, 0.5)))

stopifnot(all.equal(pvf2(50), 1))
stopifnot(all.equal(pvf2(100), 0))
stopifnot(all.equal(pvf2(75), 0.5))
stopifnot(all.equal(pvf2(c(75, 100)), c(0.5, 0)))

pvf3 <- function(x) {
  smaa.pvf(x, cutoffs=c(-0.15, 0.0, 0.25, 0.35), values=c(0, 0.1, 0.9, 1))
}
pvf4 <- function(x) {
  smaa.pvf(x, cutoffs=c(50, 75, 90, 100), values=c(1, 0.8, 0.5, 0))
}

stopifnot(all.equal(pvf3(0.35), 1.0))
stopifnot(all.equal(pvf3(-0.15), 0.0))
stopifnot(all.equal(pvf3(0.0), 0.1))
stopifnot(all.equal(pvf3(0.25), 0.9))
stopifnot(all.equal(pvf3(0.1), 2/5*0.8+0.1))

stopifnot(all.equal(pvf4(50), 1.0))
stopifnot(all.equal(pvf4(60), 1-(2/5*0.2)))
stopifnot(all.equal(pvf4(75), 0.8))
stopifnot(all.equal(pvf4(90), 0.5))
stopifnot(all.equal(pvf4(100), 0.0))
stopifnot(all.equal(pvf4(c(50,90,100)), c(1.0, 0.5, 0.0)))

pvf5 <- function(x) {
  smaa.pvf(x, cutoffs=c(50, 75, 90, 100), values=c(1, 0.8, 0.5, 0), outOfBounds="interpolate")
}

stopifnot(all.equal(pvf5(c(50,90,100)), c(1.0, 0.5, 0.0)))
stopifnot(all.equal(pvf5(c(10, 40, 101, 120)), c(1.32, 1.08, -0.05, -1)))


smaa/tests/nacw.R
library(smaa)

# First alternative beats second on all criteria
# --> NA central weights
N <- 1E4; m <- 2; n <- 3
meas <- array(dim=c(N,m,n))
meas[,1,] <- 1
meas[,2,] <- 0
pref <- matrix(1/3, nrow=N, ncol=n)

result <- smaa(meas, pref)
stopifnot(all.equal(result$cw[1,], rep(1/3, 3)))
stopifnot(all(is.na(result$cw[2,])))

ranks <- smaa.ranks(smaa.values(meas, pref))
cw <- smaa.cw(ranks, pref)
stopifnot(all.equal(cw[1,], rep(1/3, 3)))
stopifnot(all(is.na(cw[2,])))


# First alternative beats second in a single iteration
# --> R gives us a vector instead of a matrix (gee, thanks!)
meas[1,1,] <- 0
meas[1,2,] <- 1

result <- smaa(meas, pref)
stopifnot(all.equal(result$cw[1,], rep(1/3, 3)))
stopifnot(all.equal(result$cw[2,], rep(1/3, 3)))

ranks <- smaa.ranks(smaa.values(meas, pref))
cw <- smaa.cw(ranks, pref)
stopifnot(all.equal(cw[1,], rep(1/3, 3)))
stopifnot(all.equal(cw[2,], rep(1/3, 3)))


smaa/tests/rankEntropy.R
# Test the C implementation of ranking entropy (about 40x speed up)
library(smaa)

data <- dget('rankEntropy.txt')
N <- dim(data$ranks)[1]
m <- dim(data$ranks)[2]

counts <- .C("smaa_countRankings", as.integer(t(data$ranks)),
  as.integer(N), as.integer(m),
  counts=as.integer(rep(0, N)),
  NAOK=FALSE, DUP=FALSE)$counts

stopifnot(all.equal(counts[counts > 0], data$counts))
stopifnot(all.equal(smaa.entropy.ranking(data$ranks), data$entropy))


smaa/src/smaa.c
#include "smaa.h"
#include <R_ext/BLAS.h>

#include <math.h>
#include <stdlib.h>

/**
 * Rank the n-array of doubles t, writing results to r.
 * The rank of t[i] is the number of elements greater than t[i].
 * Assumes that n is small (complexity O(n^2)).
 */
static inline void smaa_rank(double const *t, int *r, int n) {
  for (int i = 0; i < n; ++i) {
    r[i] = 0;
    for (int j = 0; j < n; ++j) {
      if (t[j] > t[i]) {
        ++r[i];
      }
    }
  }
}

SEXP smaa_ranks(SEXP _v) {
  int const nAlt = nrows(_v);
  int const nIter = ncols(_v);

  _v = PROTECT(coerceVector(_v, REALSXP));
  double const *v = REAL(_v);

  SEXP _r = PROTECT(allocMatrix(INTSXP, nAlt, nIter));
  int *r = INTEGER(_r);

  for (int k = 0; k < nIter; ++k) {
    smaa_rank(v, r, nAlt);

    v += nAlt;
    r += nAlt;
  }

  UNPROTECT(2);
  return _r;
}

SEXP smaa_values(SEXP _meas, SEXP _pref, SEXP _singleWeight) {
  SEXP dim = getAttrib(_meas, R_DimSymbol);
  int const nAlt  = INTEGER(dim)[0];
  int const nCrit = INTEGER(dim)[1];
  int const nIter = INTEGER(dim)[2];
  int const singleWeight = asLogical(_singleWeight);
  const int inc1 = 1;
  const double one = 1.0, zero = 0.0; // for BLAS
  const char trans = 'N';

  _meas = PROTECT(coerceVector(_meas, REALSXP));
  _pref = PROTECT(coerceVector(_pref, REALSXP));
  double const *meas = REAL(_meas);
  double const *pref = REAL(_pref);

  SEXP _v = PROTECT(allocMatrix(REALSXP, nAlt, nIter));
  double *v = REAL(_v);

  for (int k = 0; k < nIter; ++k) {
    // calculate value of each alternative
    F77_CALL(dgemv)(&trans, &nAlt, &nCrit,
      &one, meas, &nAlt, pref, &inc1,
      &zero, v, &inc1); // t := 1Aw + 0t

    // advance measurement and weight pointers
    meas += nAlt * nCrit;
    if (!singleWeight) pref += nCrit;
    // advance alternative value pointer
    v += nAlt;
  }

  UNPROTECT(3);
  return _v;
}

SEXP smaa(SEXP _meas, SEXP _pref, SEXP _singleWeight) {
  SEXP dim = getAttrib(_meas, R_DimSymbol);
  int const nAlt  = INTEGER(dim)[0];
  int const nCrit = INTEGER(dim)[1];
  int const nIter = INTEGER(dim)[2];
  int const singleWeight = asLogical(_singleWeight);
  const int inc1 = 1;
  const double one = 1.0, zero = 0.0; // for BLAS
  const char trans = 'N';

  _meas = PROTECT(coerceVector(_meas, REALSXP));
  _pref = PROTECT(coerceVector(_pref, REALSXP));
  double const *meas = REAL(_meas);
  double const *pref = REAL(_pref);

  SEXP _h = PROTECT(allocMatrix(REALSXP, nAlt, nAlt));
  SEXP _wc = PROTECT(allocMatrix(REALSXP, nAlt, nCrit));
  Matrix h = { REAL(_h), nAlt, nAlt };
  memset(h.data, 0, nAlt * nAlt * sizeof(double));
  Matrix wc = { REAL(_wc), nAlt, nCrit };
  memset(wc.data, 0, nAlt * nCrit * sizeof(double));

  double t[nAlt];
  int r[nAlt]; // alternative ranks
  for (int k = 0; k < nIter; ++k) {
    // calculate value of each alternative
    F77_CALL(dgemv)(&trans, &nAlt, &nCrit,
      &one, meas, &nAlt, pref, &inc1,
      &zero, t, &inc1); // t := 1Aw + 0t

    // rank the alternatives
    smaa_rank(t, r, nAlt);
    for (int i = 0; i < nAlt; ++i) {
      *get(&h, i, r[i]) = *get(&h, i, r[i]) + 1; // update rank counts
      if (!singleWeight && r[i] == 0) { // update central weights
        for (int j = 0; j < nCrit; ++j) {
          *get(&wc, i, j) = *get(&wc, i, j) + pref[j];
        }
      }
    }

    // advance measurement and weight pointers
    meas += nAlt * nCrit;
    if (!singleWeight) pref += nCrit;
  }

  // normalize central weights
  if (!singleWeight) {
    for (int i = 0; i < nAlt; ++i) {
      double const r1 = *get(&h, i, 0);
      if (r1 > 0.0) {
        for (int j = 0; j < nCrit; ++j) {
          *get(&wc, i, j) = *get(&wc, i, j) / r1;
        }
      }
    }
  }

  SEXP ans = PROTECT(allocVector(VECSXP, 2));
  SET_VECTOR_ELT(ans, 0, _h);
  SET_VECTOR_ELT(ans, 1, _wc);

  UNPROTECT(5);
  return ans;
}


smaa/src/smaa.h
#include <R.h>
#include <Rinternals.h>

typedef struct Matrix {
  double * const data;
  int const nRow;
  int const nCol;
} Matrix;

/**
 * @param i Row index.
 * @param j Column index.
 */
static inline double *get(Matrix *m, int i, int j) {
  return m->data + j * (m->nRow) + i;
}

/*
 * Calculate ranks from values
 * @param _v: m * N matrix of values
 * @return m * N array of ranks
 */
SEXP smaa_ranks(SEXP _v);

/*
 * Calculate aggregate values from partial values
 * @param _meas: m * n * N array of partial values
 * @param _pref: n * N matrix of weights, or a single n-vector
 * @param _singleWeight: TRUE iff _pref is a single n-vector
 * @return m * N matrix of values
 */
SEXP smaa_values(SEXP _meas, SEXP _pref, SEXP _singleWeight);

/*
 * Calculate SMAA metrics from partial values
 * @param _meas: m * n * N array of partial values
 * @param _pref: n * N matrix of weights, or a single n-vector
 * @param _singleWeight: TRUE iff _pref is a single n-vector
 * @return A list of (1) m * m matrix of rank acceptabilities; (2) m * n matrix of central weights
 */
SEXP smaa(SEXP _meas, SEXP _pref, SEXP _singleWeight);

/*
 * Calculate piece-wise linear partial value function
 * @param x: vector of measurements (length N)
 * @param y: vector of cut-offs (length n) defining the PVF
 * @param v: vector of values corresponding to the cut-offs (length n)
 * @return vector of values (length N)
 */
SEXP smaa_pvf(SEXP _x, SEXP _y, SEXP _v);

/*
 * Count unique rankings
 * @param m * N matrix of rankings
 * @return N-vector of counts of unique rankings
 */
SEXP smaa_countRankings(SEXP _r);


smaa/src/entropyRanking.c
#include "smaa.h"

#include <stdlib.h>
#include <string.h>

// Compare 0-terminated arrays of ints
static int smaa_nAlt;
static int smaa_cmpintarr(void const *p1, void const *p2) {
  return memcmp(*(void const **)p1, *(void const **)p2, sizeof(int) * smaa_nAlt);
}

/*
 * Count unique rankings
 * @param m * N matrix of rankings
 * @return N-vector of counts of unique rankings
 */
SEXP smaa_countRankings(SEXP _r) {
  int const nAlt = nrows(_r);
  int const nIter = ncols(_r);

  _r = PROTECT(coerceVector(_r, INTSXP));
  int const *r = INTEGER(_r);

  SEXP _counts = PROTECT(allocVector(INTSXP, nIter));
  int *counts = INTEGER(_counts);
  memset(counts, 0, sizeof(int) * nIter);

  // Sort the rankings
  int const **sorted = malloc(nIter * sizeof(int *));
  for (int k = 0; k < nIter; ++k) {
    sorted[k] = r;
    r += nAlt;
  }
  smaa_nAlt = nAlt;
  qsort(sorted, nIter, sizeof(int *), smaa_cmpintarr);

  // Count the unique rankings
  int i = 0;
  ++counts[i];
  for (int k = 1; k < nIter; ++k) {
    if (memcmp(sorted[k], sorted[k-1], sizeof(int) * nAlt)) ++i;
    ++counts[i];
  }

  free(sorted);

  UNPROTECT(2);
  return _counts;
}


smaa/R/smaa.R
smaa.values <- function(meas, pref) {
  N <- dim(meas)[1]
  m <- dim(meas)[2]
  n <- dim(meas)[3]
  stopifnot(identical(dim(pref), c(N, n)) || identical(length(pref), n))

  values <- t(.Call(smaa_values, aperm(meas, c(2,3,1)), t(pref), is.vector(pref)))
  dimnames(values) <- dimnames(meas)[1:2]
  class(values) <- "smaa.values"
  values
}

plot.smaa.values <- function(x, ...) {
  boxplot(lapply(apply(x, 2, function(y) { list(y) }), function(z) { unlist(z) }),
    main="Alternative values")
}

smaa.ranks <- function(values) {
  N <- dim(values)[1]
  m <- dim(values)[2]
  ranks <- t(.Call(smaa_ranks, t(values))) + 1
  dimnames(ranks) <- dimnames(values)
  class(ranks) <- "smaa.ranks"
  ranks
}

plot.smaa.ranks <- function(x, ...) {
  plot(smaa.ra(x), ...)
}

print.smaa.ranks <- function(x, ...) {
  print(unclass(x), ...)
}

summary.smaa.ranks <- function(object, ...) {
  smaa.ra(object)
}

smaa.ra <- function(ranks) {
  N <- dim(ranks)[1]
  m <- dim(ranks)[2]

  ra <- t(apply(ranks, 2, tabulate, nbins=m) / N)
  attr(ra, "smaa.N") <- N
  class(ra) <- "smaa.ra"
  ra
}

smaa.pwi <- function(ranks) {
    nAlts <- ncol(ranks)
    nSamples <- nrow(ranks)
    grid <- expand.grid(1:nAlts, 1:nAlts)
    pwi <- matrix(apply(grid, 1, function(x) {
        sum(ranks[,x[1]] < ranks[,x[2]]) / nSamples
    }), ncol=nAlts)
    colnames(pwi) <- colnames(ranks)
    rownames(pwi) <- colnames(ranks)
    pwi
}

plot.smaa.ra <- function(x, ...) {
  barplot(t(x), main="Rank acceptabilities", ...)
}

print.smaa.ra <- function(x, ...) {
  cat(paste("Rank acceptabilities (N = ", attr(x, "smaa.N"), " iterations): \n", sep=""))
  attr(x, "smaa.N") <- NULL
  print(unclass(x), ...)
}

smaa.cw <- function(ranks, pref) {
  N <- dim(ranks)[1]
  n <- dim(pref)[2]
  stopifnot(identical(dim(pref), c(N, n)))

  cw <- t(apply(ranks, 2, function(r) {
    apply(pref[r == 1, , drop = FALSE], 2, mean)
  }))
  cw[is.nan(cw)] <- NA
  attr(cw, "smaa.N") <- N
  class(cw) <- "smaa.cw"
  cw
}

plot.smaa.cw <- function(x, ...) {
  # FIXME: use layout() instead?
  par(mar=c(8.1, 4.1, 4.1, 8.1))
  plot(NA, xlim=c(1, ncol(x)), ylim=c(0, max(x)), xlab="", ylab="Weight", xaxt='n', bty='L',
    main='Central weights', ...)
  for (i in 1:nrow(x)) {
    lines(x[i, , drop=TRUE], pch=i, type="b")
  }
  axis(side=1, at=1:ncol(x), labels=colnames(x), las=2)

  legend("topright", inset=c(-0.25,0), legend=rownames(x), pch=(1:nrow(x)), xpd=TRUE)
}

print.smaa.cw <- function(x, ...) {
  cat(paste("Central weights (N = ", attr(x, "smaa.N"), " iterations): \n", sep=""))
  attr(x, "smaa.N") <- NULL
  print(unclass(x), ...)
}

smaa.entropy.ranking <- function(ranks, p0=1) {
  N <- dim(ranks)[1]

	counts <- .Call(smaa_countRankings, t(ranks))

  p <- counts[counts > 0] / N * p0
  -sum(p * log2(p))
}

smaa.entropy.choice <- function(ra, p0=1) {
  if (class(ra) == 'smaa.ranks') { ra <- smaa.ra(ra) }
  stopifnot(class(ra) == 'smaa.ra')

  p <- ra[, 1] * p0 # first-rank acceptabilities
  p <- p[p > 0]
  -sum(p * log2(p))
}

smaa.cf <- function(meas, cw) {
  N <- dim(meas)[1]
  m <- dim(meas)[2]
  n <- dim(meas)[3]
  stopifnot(identical(dim(cw), c(m, n)))

  cf <- diag(apply(cw, 1, function(w) {
    if (all(!is.na(w))) {
      smaa.ra(smaa.ranks(smaa.values(meas, w)))[,1]
    } else {
      rep(NA, m)
    }
  }))
  names(cf) <- rownames(cw)

  result <- list(cf=cf, cw=cw)
  attr(result, "smaa.N") <- N
  class(result) <- "smaa.cf"

  result
}

print.smaa.cf <- function(x, ...) {
  cat(paste("Central weights (N = ", attr(x$cw, "smaa.N"), " iterations) and\n",
    "  confidence factors (N = ", attr(x, "smaa.N"), " iterations): \n", sep=""))
  attr(x, "smaa.N") <- NULL
  print(cbind(x$cf, x$cw), ...)
}

plot.smaa.cf <- function(x, ...) {
  plot(x$cw)
}

smaa <- function(meas, pref) {
  N <- dim(meas)[1]
  m <- dim(meas)[2]
  n <- dim(meas)[3]
  stopifnot(identical(dim(pref), c(N, n)) || identical(length(pref), n))

  result <- .Call(smaa_smaa, aperm(meas, c(2,3,1)), t(pref), is.vector(pref))
  names(result) <- c("h", "cw")

  # Introduce NAs where central weights are undefined
  cw <- t(apply(result$cw, 1, function(w) { if (sum(w) < 0.5) rep(NA, length(w)) else w }))

  rownames(cw) <- dimnames(meas)[[2]]
  colnames(cw) <- dimnames(meas)[[3]]
  class(cw) <- "smaa.cw"
  attr(cw, "smaa.N") <- N

  ra <- result$h / N
  rownames(ra) <- dimnames(meas)[[2]]
  class(ra) <- "smaa.ra"
  attr(ra, "smaa.N") <- N

  result <- list(cw=cw, ra=ra)
  class(result) <- "smaa.result"
  result
}

print.smaa.result <- function(x, ...) {
  print(x$ra)
  cat("\n")
  print(x$cw)
}

plot.smaa.result <- function(x, ...) {
  plot(x$ra, ...)
}

smaa.pvf <- function(x, cutoffs, values, outOfBounds="error") {
  stopifnot(length(cutoffs) == length(values))
  stopifnot(outOfBounds %in% c("error", "clip", "interpolate"))
  n <- length(cutoffs)
  N <- length(x)

  v <- .Call(smaa_pvf, x, cutoffs, values)

  clip <- function(v) {
    w <- v
    w[v < 0] <- 0
    w[v > 1] <- 1
    w
  }

  switch(outOfBounds,
         error=(function(v) { stopifnot(all(v >= 0) && all(v <= 1)); v })(v),
         clip=clip(v),
         interpolate=v)
}
